flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1169,6 +1136,7 @@ private void jobMasterFailed(JobID jobId, Throwable cause) {
     }
 
     CompletableFuture<Void> getJobTerminationFuture(JobID jobId) {
+        getMainThreadExecutor().assertRunningInMainThread();
```

Comment:
Same as for the `Dispatcher.localCleanup(JobID)`: I thought of having this explicit safeguard here being more helpful because it's otherwise only visible through the `DefaultResourceCleaner.cleanupAsync` JavaDoc that it should be called from the main thread.